### ISSUE TITLE
An error while runnning mnist benchmark

### ISSUE THREAD

**Issue by jyp4rk** [mc2-project/delphi#19]
Hi, benchmark does not proceed from sending and receiving keys. How can I solve the problem? 

![image](https://user-images.githubusercontent.com/37970459/132284897-362bac4d-c4a3-4cd1-a39f-566aee7b3738.png)

**Comment by ryanleh** [comment]
Hi, it looks like you're running the server on another machine -- can you confirm that your firewall allows for communication on port 8000 (the default port Delphi uses)? Also what is the output from the server?

**Comment by jyp4rk** [comment]
The output from the server
![image](https://user-images.githubusercontent.com/37970459/132424364-3dbbcca6-16fd-4f36-a28e-67570f21dd64.png)

I checked the firewall and find that the port 8000 is not opened. (I thought it open since the connection starts.) I will move the environment to AWS instance and retry. thank you.

**Comment by jyp4rk** [comment]
Everything works fine in AWS instance. I have a few questions about the paper Delphi and the code.

1. Is Batchnorm not considered in inference latency since it can be merged into other layer?
2. what is the configuration of the ResNet-32 of  3.8 seconds and 60MB in the paper? is it when num_poly is 26?
3. How can I get the online latency of using Quad-only ResNet? (replacing all the ReLUs in the network). I tried with the option --layers 32 and got 543ms but isn't it too long considering the communication and computation cost? (192× and 10000× smaller
than GC as the paper said) 

The image is the result I got using the script (for question 2) : 
server : env RAYON_NUM_THREADS=4 CLICOLOR=0 cargo +nightly run --bin resnet32-server --all-features --release -- --layers 26
client : env RAYON_NUM_THREADS=4 CLICOLOR=0 cargo +nightly run --bin resnet32-client --all-features --release -- --ip xxx.xx.xx.xxx --layers 26

![image](https://user-images.githubusercontent.com/37970459/132461557-a24f8594-ad55-4751-b0df-3eee3613aea6.png)

Thanks

**Comment by ryanleh** [comment]
1. Yes, exactly.
2. Yes, that is correct. Note that our implementation has changed a bit since the paper so I believe you should expect slightly higher communication (74.5MB vs. 60MB for the example you showed) but faster runtime. Also note that we got those numbers on the setup specified in the paper, deviating from that may give slightly different latency results. 
3. No that sounds about right. You could look at the full trace to see exactly what is taking time, but my guess is that the computational/communication cost of the activations are as small as you'd expect, but the overall runtime is bottle-necked by the linear layer computation + latency of communication.  

    Note that we were never able to train a model which obtained good accuracy with 32 quadratic layers, so, unless you were able to do this, there's no practical purpose to replacing all 32 ReLU layers.

**Comment by jyp4rk** [comment]
Thank you for answering. It helps me a lot. It seems linear layer computation + latency of communication is larger than I expected.
